typing/jkind.ml

```diff
@@ -1457,94 +1466,105 @@ module Jkind_desc = struct
         | Tlink _ | Tsubst _ ->
           Misc.fatal_error "Tlink or Tsubst in With_bounds.reduce"
     end in
-    let rec loop ctl bounds_so_far :
+    let rec loop (ctl : Loop_control.t) bounds_so_far :
         (type_expr * With_bounds_type_info.t) list ->
-        Bounds.t * with_bounds_types = function
+        Bounds.t * (l2 * r2) with_bounds * Fuel_status.t = function
       (* early cutoff *)
-      | _ when Bounds.le Bounds.max bounds_so_far ->
-        bounds_so_far, With_bounds_types.empty
-      | [] -> bounds_so_far, With_bounds_types.empty
+      | _
+        when Misc.Le_result.is_le
+               (Bounds.less_or_equal Bounds.max bounds_so_far) ->
+        (* CR layouts v2.8: we can do better by early-terminating on a per-axis basis *)
+        bounds_so_far, No_with_bounds, Sufficient_fuel
+      | [] -> bounds_so_far, No_with_bounds, ctl.fuel_status
       | (ty, ti) :: bs -> (
-        let omit (type a) (axis : a Axis.t) =
-          not (Axis_set.mem ti.relevant_axes axis)
-        in
-        let join_respecting_omit bound =
-          Bounds.Map2.f
-            { f =
-                (fun (type a) ~(axis : a Axis.t) b1 b2 ->
-                  if omit axis
-                  then b1
-                  else
-                    let (module Bound_ops) = Axis.get axis in
-                    Bound_ops.join b1 b2)
-            }
-            bounds_so_far bound
+        (* Map the type's info before expanding the type *)
+        let ti = map_type_info ty ti in
+        (* We don't care about axes that are already max because they can't get any
+           better or worse. By ignoring them, we can terminate early *)
+        let ti : With_bounds_type_info.t =
+          { relevant_axes =
+              Axis_set.diff ti.relevant_axes (Bounds.get_max_axes bounds_so_far)
+          }
         in
-        match Loop_control.check ctl ty with
-        | Stop ->
-          (* out of fuel *)
-          join_respecting_omit Bounds.max, With_bounds_types.of_list bs
-        | Skip -> loop ctl bounds_so_far bs (* skip [b] *)
-        | Continue ctl_after_unpacking_b -> (
-          match jkind_of_type ty with
-          | Some b_jkind ->
-            if (match b_jkind.quality with Not_best -> false | Best -> true)
-               || not require_best
-            then
+        match Axis_set.is_empty ti.relevant_axes with
+        | true -> loop ctl bounds_so_far bs
```

Comment:
```suggestion
        | true -> loop ctl bounds_so_far bs (* [ty] matters only for axes that are already [max]; drop from the with-bounds *)
```

Reply:
I've committed a slightly different comment that I find to be more clear